File: src/aetherius/core/runtime/drivers.py
"""DriverManager: per-run Act driver lifecycle for multi-Act composition (Jalon 2-D).

Generalises the engine's historical single ``driver = make(act)`` bind: drivers are instantiated
and set up on demand (at the first step that needs them) and torn down together at the end of the
run. The browser Acts (II/III/IV) form one inheritance chain (PhantomDriver > OracleDriver >
ContinuumDriver) designed so a higher driver runs a lower Act's steps identically — the manager
therefore *subsumes* them into a single instance of the highest browser Act the run can reach,
which is what guarantees the central invariant of composition: one browser, one stealth session,
shared by every browser step. Vector (no browser state) keeps its own driver.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..blueprint.models import Blueprint, StepModel
from ..blueprint.validator import FLOW_NESTED_FIELDS
from ..errors import ActionError

if TYPE_CHECKING:
    from .context import RunContext


IMPLEMENTED_ACTS: frozenset[str] = frozenset({"vector", "continuum", "oracle", "phantom"})

# Escalation order of the Acts; self-healing may only walk it upward (core/runtime/healing.py).
ACT_ORDER: dict[str, int] = {"vector": 0, "continuum": 1, "oracle": 2, "phantom": 3}

BROWSER_ACTS: frozenset[str] = frozenset({"continuum", "oracle", "phantom"})


def _make_driver(act: str) -> Any:
    # Drivers are imported lazily so `import aetherius` never pulls in an Act's heavy
    # dependencies (Playwright, Anthropic, ...). Each driver defers its own extra to runtime.
    if act == "vector":
        from ...acts.vector.driver import VectorDriver

        return VectorDriver()
    if act == "continuum":
        from ...acts.continuum.driver import ContinuumDriver

        return ContinuumDriver()
    if act == "oracle":
        from ...acts.oracle.driver import OracleDriver

        return OracleDriver()
    if act == "phantom":
        from ...acts.phantom.driver import PhantomDriver

        return PhantomDriver()
    raise ActionError(f"Act {act!r} is not implemented yet. Available: {sorted(IMPLEMENTED_ACTS)}.")


def collect_effective_acts(blueprint: Blueprint) -> frozenset[str]:
    """Every act a run may touch: the Blueprint act, per-step overrides (inherited through flow
    branches) and the self-healing chains a failing browser step could escalate along.

    Computed statically before the run starts so the manager can pick the single browser driver
    able to serve them all. The tree was already validated (validate_for_act walks the same shape).
    """
    acts: set[str] = {blueprint.act}
    default_chain = blueprint.options.fallback

    def visit(step: StepModel, inherited: str) -> None:
        act = step.act or inherited
        acts.add(act)
        nested_fields = FLOW_NESTED_FIELDS.get(step.action)
        if nested_fields is not None:
            for field in nested_fields:
                raw = step.extra_fields.get(field)
                if isinstance(raw, list):
                    for item in raw:
                        visit(StepModel.model_validate(item), act)
            return
        # A failing leaf may escalate along its chain — upward and between browser Acts only.
        if act in BROWSER_ACTS:
            chain = step.fallback if step.fallback is not None else default_chain
            for entry in chain:
                if ACT_ORDER.get(entry, -1) > ACT_ORDER[act]:
                    acts.add(entry)

    for step in blueprint.steps:
        visit(step, blueprint.act)
    return frozenset(acts)


class DriverManager:
    """Lazy per-run driver registry (see module docstring).

    Satisfies the executor's ``DriverResolver`` protocol (core/runtime/steps.py).
    """

    def __init__(self, blueprint: Blueprint) -> None:
        browser = [act for act in collect_effective_acts(blueprint) if act in BROWSER_ACTS]
        # The one browser driver able to serve every browser act this run can reach.
        self._browser_act: str | None = max(browser, key=ACT_ORDER.__getitem__, default=None)
        self._drivers: dict[str, Any] = {}

    def resolve_driver(self, act: str, ctx: "RunContext") -> Any:
        """Return the live driver for *act*, instantiating and setting it up on first use."""
        key = self._browser_act if act in BROWSER_ACTS and self._browser_act else act
        driver = self._drivers.get(key)
        if driver is None:
            driver = _make_driver(key)
            driver.setup(ctx)
            self._drivers[key] = driver
        return driver

    def teardown_all(self, ctx: "RunContext") -> None:
        """Tear down every live driver, newest first; the first error propagates after all ran."""
        first: Exception | None = None
        for driver in reversed(list(self._drivers.values())):
            try:
                driver.teardown(ctx)
            except Exception as exc:  # one failing teardown must not skip the others
                first = first or exc
        self._drivers.clear()
        if first is not None:
            raise first

File: sdks/engine/src/runtime/drivers.ts
/**
 * Per-run driver lifecycle, mirror of `src/aetherius/core/runtime/drivers.py`.
 *
 * Drivers are instantiated and set up on demand — at the first step that needs them — and torn
 * down together at the end of the run.
 *
 * Where Python names its four drivers in a `match`, this engine keeps a **registry**. The reason
 * is the package boundary: `@aetherius/engine` is platform-neutral, and the Continuum driver needs
 * a WebView, so it will ship in `@aetherius/react-native` (milestone 3-D) and register itself
 * there. Without the registry the neutral package would have to import a platform one, which is
 * exactly the dependency the phase set out to avoid.
 *
 * Not ported: the subsumption of the browser Acts into a single instance. It exists so Acts II/III/
 * IV share one browser; here there is one Act.
 */

import { isEmbeddedAct } from "../blueprint/capabilities.js";
import type { ActDriver, DriverFactory, HostServices, RunContext } from "../driver.js";
import { ActionError } from "../errors.js";
import { VectorDriver } from "../acts/vector/driver.js";
import type { DriverResolver } from "./steps.js";

const REGISTRY = new Map<string, DriverFactory>([
  ["vector", (host: HostServices) => new VectorDriver(host)],
]);

/**
 * Register the driver for *act*, replacing any previous one.
 *
 * The seam `@aetherius/react-native` uses to add Continuum without this package knowing it exists.
 */
export function registerDriver(act: string, factory: DriverFactory): void {
  REGISTRY.set(act, factory);
}

/** The acts a driver is registered for, in registration order. */
export function registeredActs(): readonly string[] {
  return [...REGISTRY.keys()];
}

export class DriverManager implements DriverResolver {
  private readonly drivers = new Map<string, ActDriver>();

  constructor(private readonly host: HostServices) {}

  /** The live driver for *act*, instantiated and set up on first use. */
  async resolveDriver(act: string, ctx: RunContext): Promise<ActDriver> {
    const existing = this.drivers.get(act);
    if (existing !== undefined) return existing;

    const factory = REGISTRY.get(act);
    if (factory === undefined) throw noDriver(act);

    const driver = factory(this.host);
    await driver.setup(ctx);
    this.drivers.set(act, driver);
    return driver;
  }

  /** Tear every live driver down, newest first; the first error propagates after all have run. */
  async teardownAll(ctx: RunContext): Promise<void> {
    const drivers = [...this.drivers.values()].reverse();
    this.drivers.clear();
    let first: unknown;
    for (const driver of drivers) {
      try {
        await driver.teardown(ctx);
      } catch (error) {
        // One failing teardown must not skip the others: a leaked WebView outlives the run.
        first = first ?? error;
      }
    }
    if (first !== undefined) throw first;
  }
}

/**
 * Say which of the two problems it is: an act this engine never runs (the validator already said
 * so, in better words), or one it validates but whose driver has not been brought in.
 */
function noDriver(act: string): ActionError {
  if (isEmbeddedAct(act)) {
    return new ActionError(
      `Act '${act}' is accepted by this engine but no driver is registered for it: ` +
        "import '@aetherius/react-native', which registers the WebView driver.",
    );
  }
  return new ActionError(
    `Act '${act}' is not implemented by the embedded engine. Registered acts: ` +
      `${registeredActs().join(", ")}.`,
  );
}
